ath79: allow asserting mdio/mac reset together

981213 opened the pull request:
On ar933x and later chips, there are separated mac/mdio resets, but resetting the entire gmac block with register values requires both mac_reset and mdio_reset to be asserted together.
Add support for optional mdio reset so that we can do a full reset if needed.

This PR also include several cleanups for ag71xx, check commit message for details.

Tested on:
Winchannel WB2000 (AR9344)
TP-Link TL-WDR4900 v2 (QCA9558)

For other chips, the code should behave the same as before if dts is unchanged.

981213 committed:
ath79: ag71xx: replace ag71xx_get_phy_if_mode_name() with phy_modes()

phy_modes() in phy.h can convert PHY modes to string with supports
for all available PHY modes.

Also add a space in mode printing to make it look better.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 committed:
ath79: ag71xx: replace alloc_etherdev with devm_alloc_etherdev

using the devm api makes the code simpler.
also drop unneeded memory free from ag71xx_remove since they are
allocated using devm apis.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 committed:
ath79: ag71xx-mdio: get reset control using devm api

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 committed:
ath79: ag71xx: add support for optional mdio reset

On ar933x and later chips, there are separated mac/mdio resets, but
resetting the entire gmac block with register values requires both
mac_reset and mdio_reset to be asserted together.

Add support for optional mdio reset so that we can do a full reset
if needed.

This patch also replaced deprecated devm_reset_control_get for
mac reset.

To use this feature, the following is needed:
 1. drop "simple-mfd" compatible to register mdio0 after gmac init
    so that mdio registers aren't reset after initialization.
 2. move mdio reset from mdio-bus to its parent eth node.

NOTE: This can't be applied on gmac1 with builtin switch since we
haven't add a feature to defer probe if phy connection failed.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 committed:
ath79: ar934x: assert mac/mdio reset together for gmac0

This allows resetting gmac registers during initialization.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 committed:
ath79: ag71xx: register mdio using of_platform_populate

remove the hacky checking of "simple-mfd" compatible
also add some comments explaining that piece of code.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

ynezz commented:
@981213 Thanks! I've just tested it and can confirm, that the MAC is now being properly reset:

```
[    0.847961] eth%d: mac_cfg1=80000000, mac_cfg2=00007000, ipg=40605060, hdx=00a1f037, mfl=00000600
[    0.856976] eth%d: mac_ifctl=00000000, mac_addr1=00000000, mac_addr2=00000000
[    0.864235] eth%d: fifo_cfg0=0000001f, fifo_cfg1=01ffffff, fifo_cfg2=015500aa
[    0.871491] eth%d: fifo_cfg3=015503ff, fifo_cfg4=00000000, fifo_cfg5=000bffff
```

`Tested-by: Petr Štetiar <ynezz@true.cz>`

ynezz commented:
@ae6xe Could you please test https://github.com/ynezz/openwrt/tree/upstream/nanostation-m-xw on your Nanostation M5 XW one more time and let us know if it's still working for you? Thanks!

ae6xe commented:
On Wed, Jan 9, 2019 at 4:01 PM Petr Štetiar <notifications@github.com> wrote:
>
> @ae6xe Could you please test https://github.com/ynezz/openwrt/tree/upstream/nanostation-m-xw on your Nanostation M5 XW one more time and let us know if it's still working for you? Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Successful test on NanoStation M5 XW.    laptop on main port, home
network on secondary port.  I can browse the internet on first boot.

root@OpenWrt:/# dmesg
[    0.000000] Linux version 4.14.91 (joe@AE6XE-VM) (gcc version 7.4.0
(OpenWrt GCC 7.4.0 r8978+41-eb1887be93)) #0 Wed Jan 9 23:42:11 2019
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0001974c (MIPS 74Kc)
[    0.000000] MIPS: machine is Ubiquiti Nanostation M (XW)
[    0.000000] SoC: Atheros AR9342 rev 2
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 04000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases,
linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] On node 0 totalpages: 16384
[    0.000000] free_area_init_node: node 0, pgdat 80511600,
node_mem_map 81007720
[    0.000000]   Normal zone: 128 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 16384 pages, LIFO batch:3
[    0.000000] random: get_random_bytes called from
start_kernel+0x8c/0x474 with crng_init=0
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16256
[    0.000000] Kernel command line: console=ttyS0,115200
rootfstype=squashfs,jffs2
[    0.000000] PID hash table entries: 256 (order: -2, 1024 bytes)
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Memory: 58276K/65536K available (3754K kernel code,
145K rwdata, 908K rodata, 1200K init, 204K bss, 7260K reserved, 0K
cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 51
[    0.000000] CPU clock: 535.000 MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles:
0xffffffff, max_idle_ns: 7144898866 ns
[    0.000010] sched_clock: 32 bits at 267MHz, resolution 3ns, wraps
every 8027976190ns
[    0.008326] Calibrating delay loop... 266.64 BogoMIPS (lpj=1333248)
[    0.094962] pid_max: default: 32768 minimum: 301
[    0.100113] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.107173] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.119574] clocksource: jiffies: mask: 0xffffffff max_cycles:
0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.130090] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.136678] pinctrl core: initialized pinctrl subsystem
[    0.143387] NET: Registered protocol family 16
[    0.177978] clocksource: Switched to clocksource MIPS
[    0.184610] NET: Registered protocol family 2
[    0.190190] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.197631] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.204476] TCP: Hash tables configured (established 1024 bind 1024)
[    0.211422] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.217657] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.224688] NET: Registered protocol family 1
[    0.229416] PCI: CLS 0 bytes, default 32
[    0.233268] Crashlog allocated RAM at address 0x3f00000
[    0.240452] workingset: timestamp_bits=30 max_order=14 bucket_order=0
[    0.253024] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.259290] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME)
(CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.280590] io scheduler noop registered
[    0.284770] io scheduler deadline registered (default)
[    0.292850] pinctrl-single 1804002c.pinmux: 544 pins at pa b804002c size 68
[    0.301417] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    0.309094] console [ttyS0] disabled
[    0.312946] 18020000.uart: ttyS0 at MMIO 0x18020000 (irq = 8,
base_baud = 2500000) is a 16550A
[    0.322197] console [ttyS0] enabled
[    0.329632] bootconsole [early0] disabled
[    0.349310] m25p80 spi0.0: mx25l6405d (8192 Kbytes)
[    0.354324] 5 fixed-partitions partitions found on MTD device spi0.0
[    0.360822] Creating 5 MTD partitions on "spi0.0":
[    0.365698] 0x000000000000-0x000000040000 : "u-boot"
[    0.371758] 0x000000040000-0x000000050000 : "u-boot-env"
[    0.378125] 0x000000050000-0x0000007b0000 : "firmware"
[    0.386763] 2 uimage-fw partitions found on MTD device firmware
[    0.392840] Creating 2 MTD partitions on "firmware":
[    0.397884] 0x000000000000-0x000000190000 : "kernel"
[    0.403832] 0x000000190000-0x000000760000 : "rootfs"
[    0.409791] mtd: device 4 (rootfs) set to be root filesystem
[    0.415578] 1 squashfs-split partitions found on MTD device rootfs
[    0.421923] 0x000000380000-0x000000760000 : "rootfs_data"
[    0.428365] 0x0000007b0000-0x0000007f0000 : "cfg"
[    0.434032] 0x0000007f0000-0x000000800000 : "EEPROM"
[    0.441319] libphy: Fixed MDIO Bus: probed
[    0.789933] libphy: ag71xx_mdio: probed
[    0.796800] switch0: Atheros AR8236 rev. 1 switch registered on mdio-bus.0
[    0.850500] ag71xx 19000000.eth: connected to PHY at mdio-bus.0:00
[uid=004dd043, driver=Atheros AR8216/AR8236/AR8316]
[    0.862150] eth0: Atheros AG71xx at 0xb9000000, irq 4
[    0.869982] NET: Registered protocol family 10
[    0.879586] Segment Routing with IPv6
[    0.883425] NET: Registered protocol family 17
[    0.888063] 8021q: 802.1Q VLAN Support v1.8
[    0.898818] VFS: Mounted root (squashfs filesystem) readonly on device 31:4.
[    0.913828] Freeing unused kernel memory: 1200K
[    0.918457] This architecture does not have kernel memory protection.
[    1.592045] init: Console is alive
[    1.595759] init: - watchdog -
[    2.381378] random: fast init done
[    2.489092] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    2.615475] kmodloader: done loading kernel modules from
/etc/modules-boot.d/*
[    2.633393] init: - preinit -
[    3.635053] random: jshn: uninitialized urandom read (4 bytes read)
[    3.815180] random: jshn: uninitialized urandom read (4 bytes read)
[    4.064637] random: jshn: uninitialized urandom read (4 bytes read)
[    4.952521] urandom_read: 4 callbacks suppressed
[    4.952530] random: procd: uninitialized urandom read (4 bytes read)
[    4.971081] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[    4.977091] IPv6: ADDRCONF(NETDEV_UP): eth0.1: link is not ready
[    6.019281] eth0: link up (100Mbps/Full duplex)
[    6.023974] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[    6.030790] IPv6: ADDRCONF(NETDEV_CHANGE): eth0.1: link becomes ready
[    8.312076] mount_root: jffs2 not ready yet, using temporary tmpfs overlay
[    8.347912] urandom-seed: Seed file not found (/etc/urandom.seed)
[    8.477908] procd: - early -
[    8.482101] procd: - watchdog -
[    9.125013] procd: - watchdog -
[    9.128602] procd: - ubus -
[    9.138076] eth0: link down
[    9.170573] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.182621] random: ubusd: uninitialized urandom read (4 bytes read)
[    9.190714] procd: - init -
[    9.639491] kmodloader: loading kernel modules from /etc/modules.d/*
[    9.651305] ip6_tables: (C) 2000-2006 Netfilter Core Team
[    9.670520] Loading modules backported from Linux version
v4.19.7-0-g61c68f2a2af0
[    9.678187] Backport generated by backports.git v4.19.7-1-0-g148b072d
[    9.688388] ip_tables: (C) 2000-2006 Netfilter Core Team
[    9.703910] nf_conntrack version 0.5.0 (1024 buckets, 4096 max)
[    9.781335] xt_time: kernel timezone is -0000
[    9.858912] PPP generic driver version 2.4.2
[    9.866213] NET: Registered protocol family 24
[    9.921290] ath: EEPROM regdomain: 0x2a
[    9.921300] ath: EEPROM indicates we should expect a direct regpair map
[    9.921321] ath: Country alpha2 being used: US
[    9.921326] ath: Regpair used: 0x2a
[    9.937067] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[    9.939291] ieee80211 phy0: Atheros AR9340 Rev:2 mem=0xb8100000, irq=11
[   10.040004] kmodloader: done loading kernel modules from /etc/modules.d/*
[   21.497384] jffs2_scan_eraseblock(): End of filesystem marker found at 0x0
[   21.541270] jffs2_build_filesystem(): unlocking the mtd device...
[   21.541283] done.
[   21.549569] jffs2_build_filesystem(): erasing all blocks after the
end marker...
[   27.701309] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[   27.771565] br-lan: port 1(eth0.1) entered blocking state
[   27.777062] br-lan: port 1(eth0.1) entered disabled state
[   27.783027] device eth0.1 entered promiscuous mode
[   27.787900] device eth0 entered promiscuous mode
[   27.915178] IPv6: ADDRCONF(NETDEV_UP): br-lan: link is not ready
[   28.031292] IPv6: ADDRCONF(NETDEV_UP): eth0.2: link is not ready
[   28.739703] eth0: link up (100Mbps/Full duplex)
[   28.744385] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   28.818148] br-lan: port 1(eth0.1) entered blocking state
[   28.823647] br-lan: port 1(eth0.1) entered forwarding state
[   28.829720] IPv6: ADDRCONF(NETDEV_CHANGE): eth0.2: link becomes ready
[   28.929354] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   50.310506] done.
[   50.312518] jffs2: notice: (988) jffs2_build_xattr_subsystem:
complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan)
and 0 of xref (0.
[   50.515861] overlayfs: upper fs does not support tmpfile.
[  122.928036] random: crng init done

root@OpenWrt:/# ifconfig
br-lan    Link encap:Ethernet  HWaddr 04:18:D6:A7:70:B5
          inet addr:192.168.1.1  Bcast:192.168.1.255  Mask:255.255.255.0
          inet6 addr: fe80::618:d6ff:fea7:70b5/64 Scope:Link
          inet6 addr: fdf7:61cd:992a::1/60 Scope:Global
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:4623 errors:0 dropped:0 overruns:0 frame:0
          TX packets:6043 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:544816 (532.0 KiB)  TX bytes:6610191 (6.3 MiB)

eth0      Link encap:Ethernet  HWaddr 04:18:D6:A7:70:B5
          inet6 addr: fe80::618:d6ff:fea7:70b5/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:10815 errors:0 dropped:0 overruns:0 frame:0
          TX packets:10657 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:7279578 (6.9 MiB)  TX bytes:7255548 (6.9 MiB)
          Interrupt:4

eth0.1    Link encap:Ethernet  HWaddr 04:18:D6:A7:70:B5
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:4623 errors:0 dropped:0 overruns:0 frame:0
          TX packets:6043 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:544816 (532.0 KiB)  TX bytes:6610191 (6.3 MiB)

eth0.2    Link encap:Ethernet  HWaddr 04:18:D6:A7:70:B5
          inet addr:172.16.0.58  Bcast:172.16.0.255  Mask:255.255.255.0
          inet6 addr: fe80::618:d6ff:fea7:70b5/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:6192 errors:0 dropped:0 overruns:0 frame:0
          TX packets:4594 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:6540092 (6.2 MiB)  TX bytes:600837 (586.7 KiB)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:88 errors:0 dropped:0 overruns:0 frame:0
          TX packets:88 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:6517 (6.3 KiB)  TX bytes:6517 (6.3 KiB)
root@OpenWrt:/# swconfig dev switch0 show
Global attributes:
        enable_vlan: 1
        enable_mirror_rx: 0
        enable_mirror_tx: 0
        mirror_monitor_port: 0
        mirror_source_port: 0
        arl_table: address resolution table
Port 0: MAC 04:18:d6:a7:70:b5
Port 1: MAC a4:77:33:4d:e6:3a
Port 1: MAC 00:21:55:9a:75:c0
Port 1: MAC 04:c0:5b:80:24:3e
Port 1: MAC 00:00:54:fe:07:cc
Port 1: MAC dc:9f:db:54:26:a0
Port 1: MAC 20:e5:2a:09:ef:84
Port 1: MAC 54:33:cb:67:c4:56
Port 5: MAC 5c:26:0a:05:3c:62

Port 0:
        mib: MIB counters
RxBroad     : 34
RxPause     : 0
RxMulti     : 53
RxFcsErr    : 0
RxAlignErr  : 0
RxRunt      : 0
RxFragment  : 0
Rx64Byte    : 33
Rx128Byte   : 5223
Rx256Byte   : 452
Rx512Byte   : 388
Rx1024Byte  : 259
Rx1518Byte  : 4332
RxMaxByte   : 7
RxTooLong   : 0
RxGoodByte  : 7301632 (6.9 MiB)
RxBadByte   : 0
RxOverFlow  : 0
Filtered    : 56
TxBroad     : 128
TxPause     : 0
TxMulti     : 188
TxUnderRun  : 0
Tx64Byte    : 0
Tx128Byte   : 5368
Tx256Byte   : 534
Tx512Byte   : 404
Tx1024Byte  : 264
Tx1518Byte  : 4341
TxMaxByte   : 9
TxOverSize  : 0
TxByte      : 7352441 (7.0 MiB)
TxCollision : 0
TxAbortCol  : 0
TxMultiCol  : 0
TxSingleCol : 0
TxExcDefer  : 0
TxDefer     : 0
TxLateCol   : 0

        pvid: 0
        link: port:0 link:up speed:100baseT full-duplex
Port 1:
        mib: MIB counters
RxBroad     : 91
RxPause     : 0
RxMulti     : 59
RxFcsErr    : 0
RxAlignErr  : 0
RxRunt      : 0
RxFragment  : 0
Rx64Byte    : 16
Rx128Byte   : 1263
Rx256Byte   : 321
Rx512Byte   : 235
Rx1024Byte  : 137
Rx1518Byte  : 4250
RxMaxByte   : 0
RxTooLong   : 0
RxGoodByte  : 6654347 (6.3 MiB)
RxBadByte   : 0
RxOverFlow  : 0
Filtered    : 0
TxBroad     : 3
TxPause     : 0
TxMulti     : 4
TxUnderRun  : 0
Tx64Byte    : 14
Tx128Byte   : 4081
Tx256Byte   : 146
Tx512Byte   : 123
Tx1024Byte  : 120
Tx1518Byte  : 88
TxMaxByte   : 0
TxOverSize  : 0
TxByte      : 610001 (595.7 KiB)
TxCollision : 0
TxAbortCol  : 0
TxMultiCol  : 0
TxSingleCol : 0
TxExcDefer  : 0
TxDefer     : 0
TxLateCol   : 0

        pvid: 2
        link: port:1 link:up speed:100baseT full-duplex auto
Port 2:
        mib: No MIB data
        pvid: 0
        link: port:2 link:down
Port 3:
        mib: No MIB data
        pvid: 0
        link: port:3 link:down
Port 4:
        mib: No MIB data
        pvid: 0
        link: port:4 link:down
Port 5:
        mib: MIB counters
RxBroad     : 458
RxPause     : 0
RxMulti     : 129
RxFcsErr    : 0
RxAlignErr  : 0
RxRunt      : 0
RxFragment  : 0
Rx64Byte    : 18
Rx128Byte   : 4143
Rx256Byte   : 239
Rx512Byte   : 168
Rx1024Byte  : 141
Rx1518Byte  : 379
RxMaxByte   : 31
RxTooLong   : 0
RxGoodByte  : 1137112 (1.0 MiB)
RxBadByte   : 0
RxOverFlow  : 0
Filtered    : 421
TxBroad     : 5
TxPause     : 0
TxMulti     : 19
TxUnderRun  : 0
Tx64Byte    : 19
Tx128Byte   : 1129
Tx256Byte   : 300
Tx512Byte   : 233
Tx1024Byte  : 135
Tx1518Byte  : 4250
TxMaxByte   : 0
TxOverSize  : 0
TxByte      : 6636645 (6.3 MiB)
TxCollision : 0
TxAbortCol  : 0
TxMultiCol  : 0
TxSingleCol : 0
TxExcDefer  : 0
TxDefer     : 0
TxLateCol   : 0

        pvid: 1
        link: port:5 link:up speed:100baseT full-duplex auto
VLAN 1:
        vid: 1
        ports: 0t 5
VLAN 2:
        vid: 2
        ports: 0t 1

981213 committed:
ath79: qca955x: assert mdio/gmac reset together

This allows resetting gmac registers during initialization.
Also add compatible string for qca955x mdio to enable more mdio
clock dividers.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

chunkeey commented on target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c:
```diff
@@ -1540,9 +1520,8 @@ static int ag71xx_probe(struct platform_device *pdev)
 
-	pr_info("%s: Atheros AG71xx at 0x%08lx, irq %d, mode:%s\n",
-		dev->name, (unsigned long) ag->mac_base, dev->irq,
-		ag71xx_get_phy_if_mode_name(ag->phy_if_mode));
```
if you want, ag71xx_get_phy_if_mode_name() could also be replaced by: [phy_modes()](https://elixir.bootlin.com/linux/v4.14.96/source/include/linux/phy.h#L113)

The phy mode string will be in lower case, but that should be the only downside.

981213 replied:
That's pretty good. I've changed to use that.

neheb commented:
I've tested this on an Archer C7v2. Works fine.

dangowrt commented:
applied to master
https://github.com/openwrt/openwrt/commit/47f0be676fb700a8ab071fd0fe6f5d818548a852
https://github.com/openwrt/openwrt/commit/51cdf0e9eb15d2489a36b705bb32856f80857031
https://github.com/openwrt/openwrt/commit/cea3e9e4383688f00ce5ef76c2b694aeead4c0ac
https://github.com/openwrt/openwrt/commit/596a5325cd7255faa94aa5e13ae97a6ee64f0bd2
https://github.com/openwrt/openwrt/commit/f92bbdcc93d9c25d14112da57c5aa63169a4840f
https://github.com/openwrt/openwrt/commit/a4eef43a120d51bb5386730b820704e5cb5e1acf
https://github.com/openwrt/openwrt/commit/f73b2d64ed56c1ba41c1ef906ef292d07f839def
